### PR TITLE
Add retry when using async method to perform request in AsyncHttpConnector

### PR DESCRIPTION
**Pull request by zymap**
---

Master Issue: #<xyz>

*Motivation*

There are two methods in AsycnHttpConnector to perform a request
for pulsar admin. One is the sync method and another is the async
method. We only add the multi-host logic in the sync method,
this PR makes the async method can handle the multi-host URL.

*Modifications*

- Make the async apply method in the `AsycnHttpConnector` can handle the multi-host

### CODE REVIEW

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java [hunk 1]
```diff
@@ -176,39 +153,83 @@ private URI replaceWithNew(InetSocketAddress address, URI uri) {
         return URI.create(newUri);
     }
 
+    @Override
+    public Future<?> apply(ClientRequest jerseyRequest, AsyncConnectorCallback callback) {
+        final CompletableFuture<ClientResponse> resp = new CompletableFuture<>();
+
 
+        CompletableFuture.runAsync(() -> {
```
Comment (sijie) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java]: I think we should rewrite this logic in async way. It seems that we try to wrap a sync code in an async way.

**Comment by tuteng** [comment]
@zymap Please look at the comment of sijie.

**Comment by zymap** [comment]
@sijie PTAL. Thanks.

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java [hunk 2]
```diff
@@ -176,39 +153,39 @@ private URI replaceWithNew(InetSocketAddress address, URI uri) {
         return URI.create(newUri);
     }
 
+    @Override
+    public Future<?> apply(ClientRequest jerseyRequest, AsyncConnectorCallback callback) {
+        final CompletableFuture<ClientResponse> respFuture = new CompletableFuture<>();
+        triedAddressees = new HashSet<>();
```
Comment (sijie) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java]: Why do we need `triedAddresses` to be a member variable? This method will be used concurrently.

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java [hunk 3]
```diff
@@ -176,39 +153,39 @@ private URI replaceWithNew(InetSocketAddress address, URI uri) {
...
-            result = resultFuture.get(timeout, TimeUnit.MILLISECONDS);
-        } else {
-            result = resultFuture.get();
-        }
-
-        if (result != null && result instanceof Throwable) {
-            throw new ExecutionException((Throwable) result);
+    private void doNextRetry(ClientRequest jerseyRequest, AsyncConnectorCallback callback, CompletableFuture<ClientResponse> respFuture) {
+        InetSocketAddress nextTriedAddress = serviceNameResolver.resolveHost();
+        log.info("Trying to use address [{}] to send request.", nextTriedAddress.toString());
```
Comment (sijie) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java]: why is this logging level `info`?

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java [hunk 4]
```diff
@@ -176,39 +153,39 @@ private URI replaceWithNew(InetSocketAddress address, URI uri) {
...
+        if (triedAddressees.contains(nextTriedAddress)) {
+            Exception triedAllException = new ProcessingException("All addresses are tried and failed");
```
Comment (sijie) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java]: I don't think we should use this as the condition to exist. We need to check if we have tried at least multiple times, right?

**Comment by jiazhai** [comment]
@zymap for the new comments

**Comment by zymap** [comment]
Thanks for reminder, I am trying to resolve it. This PR is caused some tests to be flaky.

**Comment by zymap** [comment]
This problem has fixed in the #6547 , close it.

**Comment by tuteng** [comment]
This pull request is closed, so remove label release-2.5.1

**Pull request closed without merging**